lib/Apache2/AuthCookieDBI.pm

```diff
@@ -557,15 +551,15 @@ sub _percent_decode {
 # _dbi_connect -- Get a database handle.
 
 sub _dbi_connect {
-    my ( $class, $r ) = @_;
+    my ( $class, $r, $config_hash ) = @_;
     Carp::confess('Failed to pass Apache request object') if not $r;
 
     my ( $pkg, $file, $line, $sub ) = caller(1);
     my $info_message = "${class}\t_dbi_connect called in $sub at line $line";
     $class->logger( $r, Apache2::Const::LOG_INFO, $info_message, undef,
         LOG_TYPE_SYSTEM, $r->uri );
 
-    my %c = $class->_dbi_config_vars($r);
+    my %c = $config_hash ? %$config_hash : $class->_dbi_config_vars($r);
```

Comment:
In this method, and all private methods, that use the config hash, perhaps we do not need to check if `$config_hash` is true because all the callers of this method are in this class, so we do not need to guard against someone calling this method without passing `$config_hash` so maybe we should just do: `my %c = %$config_hash;` - what do you think?

Reply:
If it were a new method, sure, but I worry that someone out there has subclassed AuthCookieDBI. If so, they most likely have existing calls to `_dbi_connect()`, regardless of whether you consider it to be private or not, just as I did in `Apache2_4::AuthCookieDBI` when I developed that. Perl doesn’t enforce it, and we’re not using Moose or the like here. So I recommend keeping `$config_hash` as optional.

As for tests, I’ve tried and failed. I can’t get any Apache 2.4.x tests to work, even a simple `use_ok()`. I get “bareword” errors for the Apache 2.4.x AUTHZ_* constants. I think my server’s non-chrooted environment is messed up from too many Apache and mod_perl upgrades. It’s rather complicated, and I don’t have the permissions to change things. I could maybe try testing inside a Docker container, but I’d have to look around for one with Apache 2.4.41 and mod_perl 2.0.11....

Also, I think adding Apache 2.4.x-specific tests might change the dependency requirements on Apache::Test and maybe Apache itself. IIRC, Apache::Test’s minimum Apache version check requires Apache 2.0.40.